src/dftbp/common/file.F90

```diff
@@ -5,114 +5,459 @@
+
+    type(TOpenOptions) :: opts
+    integer :: ioStat_
+    character(1024) :: ioMsg_
 
     if (present(options)) then
       opts = options
     end if
-    allocate(this)
+
+    if (present(mode)) then
+      call opts%setMode(mode, ioStat_, ioMsg)
+      if (ioStat_ /= 0) then
+        if (present(ioStat)) then
+          ioStat = ioStat_
+          return
+        else
+          call error("Invalid mode specification '" // trim(mode) // "'")
+        end if
+      end if
+    end if
+
+    if (opts%access == "default") then
+      select case (opts%action)
+      case ("read")
+        opts%access = defaultFileAccess(1)
+      case ("write")
+        opts%access = defaultFileAccess(2)
+      case ("readwrite")
+        opts%access = defaultFileAccess(3)
+      end select
+    end if
+
+    if (opts%status == "oldnew") then
+      if (fileExists(file)) then
+        opts%status = "old"
+      else
+        opts%status = "new"
+      end if
+    end if
+
+    open(newunit=this%unit, file=file, access=opts%access, action=opts%action, form=opts%form,&
+        & status=opts%status, position=opts%position, iostat=ioStat_, iomsg=ioMsg_)
+
+    if (ioStat_ /= 0) then
+      this%unit = -1
+      if (present(ioStat) .or. present(ioMsg)) then
+        if (present(ioStat)) then
+          ioStat = ioStat_
+        end if
+        if (present(ioMsg)) then
+          ioMsg = trim(ioMsg_)
+        end if
+        return
+      else
+        call error("Failed to open file '" // trim(file) // "'")
+      end if
+    end if
     if (present(ioStat)) then
-      open(newunit=this%unit, file=name, access=opts%access, action=opts%action, form=opts%form,&
-          & status=opts%status, iostat=ioStat)
-      if (ioStat /= 0) deallocate(this)
-    else
-      open(newunit=this%unit, file=name, access=opts%access, action=opts%action, form=opts%form,&
-          & status=opts%status)
+      ioStat = 0
     end if
+    this%file = trim(file)
+    this%needsClosing_ = .true.
+
+  end subroutine TFileDescr_connectToFile
+
 
-  end subroutine TFile_create
+  !> Connects the descriptor to an existing file unit
+  subroutine TFileDescr_connectToUnit(this, unit)
+
+    !> File descriptor on exit
+    class(TFileDescr), intent(out) :: this
+
+    !> Unit number to wrap as file (unit will NOT be closed on finalization)
+    integer, intent(in) :: unit
+
+    this%unit = unit
+    this%file = ""
+
+  end subroutine TFileDescr_connectToUnit
 
 
   !> Finalizes (closes) a file
-  subroutine TFile_final(this)
+  elemental impure subroutine TFileDescr_disconnect(this)
 
     !> Instance
-    type(TFile), intent(inout) :: this
+    class(TFileDescr), intent(inout) :: this
 
-    if (this%unit /= -1) then
+    if (this%unit /= -1 .and. this%needsClosing_) then
       close(this%unit)
     end if
+    this%unit = -1
+    if (allocated(this%file)) then
+      deallocate(this%file)
+    end if
+    this%needsClosing_ = .false.
```

Comment:
For files, opened via `%connectToFile()` or `openFile()` they are equivalent. However, one can also wrap a unit, with `%connectToUnit()` which had been opened already otherwise. We use this on the Fortran side of the C-API, where depending on the input, we either open a file ourselves, or use `output_unit` which is already open by default. (See https://github.com/aradi/dftbplus/blob/4ab9896810788816ab318f621f66ab8a8ef2d552/src/dftbp/api/mm/capi.F90#L517-L532)